### DISCUSSION TITLE
Laravel DBs to store dialog between requests

### DISCUSSION THREAD

**Discussion by dimaodnokoz** [koot-labs/telegram-bot-dialogs#21] [Ideas]
Hello. Thank you for maintaining the package and last update. I explored it in details. I use virtual hosting for chatbot scripts and Redis is separate paid service there. As i checked, the same situation with other virtual hosting providers and only starting from VPS Redis is included. But it is more expensive to have VPS. So i think it will be good to have option to use for example mySQL to store dialogs, as mySQL is free everywhere. I understand, that Radis is more quick because of its architecture etc, but in the majority of my projects it's not so important. I suppose i'm not the one here.

As i understand (i'm new to laravel) laravel uses DB facade to give common interface to work with different data base types. I tried to implement the use of it for task above taking code for Redis in this package as example. Changes where made to files: Laravel/DialogServiceProvider.php, Laravel/config/telegramdialogs.php. And file Laravel/Stores/DatabaseStoreAdapter.php was added. I implemented it using DB facade to make it universal in choose of db type. But i did not add Laravel/Storages/Drivers/DatabaseStore.php to use without framework as Laravel/Storages/Drivers/RedisStore.php, because i'm just not so experienced yet to do that for all db types at once.

I'd like to use this package, you kindly maintain (not separate my own). But i'm new to github and do not understand how to test my implementation and how to correctly make PR later. Probably we may work together a little to enhance this package with my implementation for DBs? Or i just give you my files (it seems not too hard for experienced dev)? Or maybe you may advice something here.

**Comment by dimaodnokoz** [comment]
telegramdialogs:
https://gist.github.com/dimaodnokoz/391c0ceabaf917b001fc7f07ae4ce5cb

DialogServiceProvider:
https://gist.github.com/dimaodnokoz/da9a043864032feb553d3f3e3baf2717

DatabaseStoreAdapter:
https://gist.github.com/dimaodnokoz/0145e4bd138fb446b1a480e2cb7bcdcb

**Reply by alies-dev** [reply]
Hey @dimaodnokoz 

Thank you for your detailed answer. I found that Store/Storage functionality is out of the responsibility of this package and added support of PSR-16 storages in one of the latest releases. I also updated installation instructions here: https://github.com/koot-labs/telegram-bot-dialogs/blob/master/docs/using-without-framework.md

So, in v1.0 of this package, I'm going to remove any code related to storage management. PSR-16 is an interface that allows to use any implementation, including:
 - Redis
 - File
 - Memory
 - SQL

However, I haven't properly tested it installation using this manual. I would appreciate of someone will do it. Also, unfortunately, I don't have enough time right now to help anyone to teach how to use GitHub to properly create PRs

**Reply by alies-dev** [reply]
Regarding Redis:

there are several hosting services that offer free tiers or free trials of Redis or Redis-like databases. Here are some popular options:

Redis Labs (Redis Cloud): Redis Labs provides a free tier of their managed Redis service, called Redis Cloud. The free tier usually includes a small amount of memory and limited features but is sufficient for development and small applications.

Heroku: Heroku offers a variety of add-ons for databases, including Redis. The Heroku Redis add-on has a free tier that provides a limited amount of memory and connections. It's great for development and testing environments.

Amazon Web Services (AWS): AWS offers a free tier for its ElastiCache service, which includes Redis. The free tier is typically available for 12 months and provides a managed Redis instance with limited resources.

Google Cloud Platform (GCP): GCP offers Memorystore, which supports Redis. While there's no permanent free tier like some other services, Google Cloud often provides credits for new users that can be used to experiment with various services, including Redis.

Microsoft Azure: Azure provides a managed Redis service through Azure Cache for Redis. Like GCP, there's no permanent free tier, but new users can get credits that may be applied toward using Azure Cache for Redis.

Fly.io: This platform offers a free tier for running small applications, including Redis instances. It's more flexible but requires more setup compared to fully managed Redis services.

**Comment by dimaodnokoz** [comment]
@alies-dev, thank you for answer about hosting providers with Redis support. But i have reasons to keep using my hosting provider and will not use Redis there. So i'm still interested in other Store/Storage variants. I'm afraid to misunderstand you, so please, let's clarify you answer about Store/Storage and PSR-16.

1) As i understood in v1.0 you will delete at all:
- src/Laravel/config/telegramdialogs.php
- src/Laravel/DialogsServiceProvider.php
- everything in src/Laravel/Stores
- src/Storages

Instead of that everyone should just install necessary storage from PSR-16 and create it like $store = new Psr16Cache... everytime in webhook to initialize $dialogManager and then work with dialog? I mean just php, no laravel facades etc like it is now?

2) How do you estimate a time of v1.0 release? I need to use dialogs in my new project and i'm afraid of doing mistakes because of misunderstanding, which may fail future updates from you at my side. So i need to understand the time perspective and decide what to do with it (wait for v1.0 release, or make my own small package from all i understand from yours as it is now).

3) Why don't you choose Laravel's caching system (config/cache.php) and Laravel's Cache facade? It seems it may be very similar in use. To make package useful with and without framework, right?

Sorry for probably silly questions, i'm just getting acquainted with all that and have no full picture yet. And Thank you very much for your work with this package. It is very useful for bot-projects.

**Reply by alies-dev** [reply]
ok, let me elaborate.

One of the challenges of this package is that it's developed to work without frameworks (even without Laravel) and with Laravel. We should consider both cases.
1. With Laravel: Laravel already provides PSR-16 implementation out of the box, it's `\Illuminate\Contracts\Config\Repository` interface and their implementations. E.g. `app('cache')` returns `CacheManager`, `app("cache")->store("redis")` returns `Illuminate\Cache\Repository`. So, in v1.0,  code in ServiceProvide will look like (an oversimplified version):
```php
$this->app->when(\KootLabs\TelegramBotDialogs\DialogManager::class)
    ->needs(\Psr\SimpleCache\CacheInterface::class)
    ->give(function (): \Illuminate\Contracts\Cache\Repository {
        $repository = app('cache')->store(config('telegramdialogs.cache_store'));
        if (! empty(config('telegramdialogs.cache_tag'))) {
            return $repository->tags(config('telegramdialogs.cache_tag'));
        }

        return $repository;
    });
```
2. For non-Laravel setups, a developer needs to instantiate PSR-16 compatible Cache implementation manually.

Giving this, only classes marked [here](https://github.com/koot-labs/telegram-bot-dialogs/commit/61af50a1af36acf2c42a62d88082fc0f7936b441) as deprecated will be removed.



>2. How do you estimate a time of v1.0 release?

I would recommend you not wait for this release, because this is a project I work on in my spare time and I don't have enough time to implement all my ideas I wanted to implement for 1.0. Plus, there are not any other active contributors to the package that will help me with that.


> 3. Why don't you choose Laravel's caching system (config/cache.php) and Laravel's Cache facade? It seems it may be very similar in use. To make package useful with and without framework, right?

That was exactly my plan and why I moved to PSR-16 as it works both with Laravel and without it.

**Reply by alies-dev** [reply]
UPD: I found some time today and created a PR (unfinished) to reuse Laravel Cache: https://github.com/koot-labs/telegram-bot-dialogs/pull/22

The only missing part is key prefixes and auto tests. Probably I'll finish it on during this month. But you can help me by testing it

**Comment by dimaodnokoz** [comment]
@alies-dev thank you for your involvement. I'd like to help with tests and even contribute, but i'm not so experienced here. Most of things you show are new for me and i spent hours to explore and study them. It's cool, that i have like a directions from you, but mostly my involvement looks like "i know some bricks, but still can not build the house".

During last week everything totally changed for me here 3 times). Each time i like start studying from zero to catch your updates. I can not keep up with you. I'd like to use your package, because i see how you enhance it. Basically i need only dialogs working now. But for now i do not understand what and how should i use in this package in my project to keep it alive after your v1.0 release. I'm playing with 

$store = new Psr16Cache(new PdoAdapter(DB::connection()->getPdo(), 'tg:dialogs::', 0));

As i understood it will store data in mysql db. Also it looks like your approach in docs and new PR. But i still didn't make it to work. I keep trying, maybe i will understand more, diving deeper in everything described by you above.

**Reply by alies-dev** [reply]
oh, that's a red flag for documentation if it's still not clear how to use the package. I know there is a gap in routing/handling Telegram updates, but it's not the responsibility of this package. Telegram Bot API PHP SDK also doesn't really solve this problem: it's too focus on bot commands (like `/start`). I think this part is the most confusing for new comers. (but I can be wrong here). I also found some bugs in old code examples and fixed them yesterday.


also, please do not worry about the 1.0 version: biggest breaking changes will be introduced by https://github.com/koot-labs/telegram-bot-dialogs/pull/22, but if you use PSR-16 store and do not use functionality marker as `@internal`, `@deprecated` or `experimental` - you are in a safe zone. The biggest challenge you may have: use for find-and-replace to migrate to a new package version. But again, I'm not going to invest sufficient time into this package next months. My intention is to make it more friendly for the community now and then accept PRs from the community to improve the package.  That's why I'm motivated to help you and other package users, especially help to potential contributors.


```php
$store = new Psr16Cache(new PdoAdapter(DB::connection()->getPdo(), 'tg:dialogs::', 0));
```

is it Laravel code (`DB` class)? If so, it's too complex. You just need something like this: https://github.com/koot-labs/telegram-bot-dialogs/pull/22/files#diff-bcdbc9b1520a1c5e2906c5e1834afee8cf206f6f97e1c233a947f28436497f3cR41
